### DISCUSSION TITLE
How to get log files on minor error when using tm1.cells.write with use_ti=True?

### DISCUSSION THREAD

**Discussion by autoneurino** [cubewise-code/tm1py#878] [Q&A] [answered]
Hello, simple question, is there a way to get the status & log file when running tm1.cells.write(use_ti=True)? I am asking for the case where you would have minor errors. Below a small sample case:

```
with TM1Service(**tm1_cloud) as tm1:   
    cells = {("not-existing", "not-existing"): -1}    
    try:
         tm1.cells.write("System Info", cells, use_ti=True)
    except:
         pass # retrieve log files here
```

**Comment by MariusWirtz** [comment, marked as answer]
You can catch the `TM1pyWritePartialFailureException` exception and access the `error_log_files` property.
Now that you know the names of the error log files, you can use the `get_error_log_file_content` function to retrieve the content of the error log file.

``` python
with TM1Service(**tm1_cloud) as tm1:
    cells = {
        ("not-existing", "not-existing"): -1
    }
    try:
        tm1.cells.write("System Info", cells, use_ti=True)

    except TM1pyWritePartialFailureException as ex:
        files = ex.error_log_files
        for file_name in files:
            error_log_content = tm1.processes.get_error_log_file_content(file_name)
            print(error_log_content)
```

This will provide you with the information you need:

```
﻿Error: Prolog procedure line (3): Invalid key: Dimension Name: "System Info Parameter", Element Name (Key): "not-existing"
```

**Reply by Kevin-Dekker** [reply]
Note that admin rights are required to use the use_ti=True argument. In case your user doesn't have admin rights, you could consider writing the cells one by one and handling exceptions:

```python
 cells = {
    ("not-existing", "not-existing"): -1,
    ("not-existing-either", "not-existing-either"): -1
}

for k, v in cells.items():
    try:
        tm1.cells.write("System Info", {k: v}) # writes a single value and returns an exception in case of failure.
    except Exception as e:
        logging.exception(e)
        print(e) # or do what you want to do in case of exception
```

This approach allows minor errors when writing cells to a cube without admin rights. 

@MariusWirtz, is there a more efficient way for larger datasets (/cellsets) where this occurs or are getting admin rights the best alternative here?

**Reply by Kevin-Dekker** [reply]
In case the error comes from non-existing (n-level) dimension elements, you could check the existence of all elements prior to the write and filter on cellsets with existing dimension elements. This will not get you tm1 logs and on failure, the whole write will fail. 

```python
    with TM1Service(**tm1_cloud) as tm1:
        cellset = {
            ("not-existing", "not-existing"): -1
        }
        cube_name = 'My Cube'
        
        # retrieves all dimension names in the cube
        dims = tm1.cubes.get(cube_name).dimensions  
        
        # retrieves all dimension elements for each dimension
        dim_elements = [tm1.elements.get_leaf_element_names(dim, dim) for dim in dims]  
        
        # creates a cellset made up of key, value pairs of which all keys can be found in their respective dimensions.
        cells_possible_to_write = {k: v for k, v in cellset.items() if all([k[_] in dim_elements[_] for _ in range(len(dims))])} 

        # creates a cellset made up of key, value pairs of which at least one of the keys cannot be found in their respective dimensions.
        cell_not_possible_to_write = {k: v for k, v in cellset.items() if not all([k[_] in dim_elements[_] for _ in range(len(dims))])}  
        
        # write cells to tm1 that have elements existing in their dimensions
        tm1.cells.write(cube_name, cellset_as_dict=cells_possible_to_write)  

        # log those cases where at least one element does not exist in the respective dimension.
        logging.warning(f'Some cells were not written to tm1. \n {cell_not_possible_to_write}')  

```

**Reply by MariusWirtz** [reply]
Actually, that's not a reliable solution for two reasons.

- the element names are case and space insensitive
- aliases can be used in the write function instead of element names

if you use the `get_all_leaf_element_identifiers` function instead of `get_leaf_element_names` it should work!

**Reply by Kevin-Dekker** [reply]
Thanks Marius, I wasn't aware of that function. What you mention was exactly the problem we faced afterwards. We retrieved the alias based on the element names as a step inbetween. Does `get_all_leaf_element_identifiers` also solve the case and space sensitivity?

**Reply by MariusWirtz** [reply]
Yes. The `get_all_leaf_element_identifiers` returns a `CaseAndSpaceInsensitiveSet` of all leaf element names and aliases. We built it exactly for the purpose to verify write operations before they happen. Sorry for the delayed reply on this subject.

**Reply by MariusWirtz** [reply]
on a side note, you might want `dim_elements` generally to be a `set`. It could make a large performance impact on a large dimension. The set should offer a near-instant contains check

**Reply by Kevin-Dekker** [reply]
Thanks! A list of sets (CaseAndSpaceInsensitiveSet) in this case . 

For completeness, here the full sample:
```python
with TM1Service(**tm1_cloud) as tm1:
    cellset = {(not-existing", "not-existing"): -1}
    cube_name = 'My Cube'
        
    # retrieves all dimension names in the cube
    dims = tm1.cubes.get(cube_name).dimensions  
        
    # retrieves all dimension elements for each dimension
    dim_elements = [tm1.elements.get_all_leaf_element_identifiers(dim, dim) for dim in dims]

    # creates a cellset made up of key, value pairs of which all keys can be found in their respective dimensions.
    cells_possible_to_write = {k: v for k, v in cellset.items() if all([k[_] in dim_elements[_] for _ in range(len(dims))])} 

    # creates a cellset made up of key, value pairs of which at least one of the keys cannot be found in their respective dimensions.
    cell_not_possible_to_write = {k: v for k, v in cellset.items() if not all([k[_] in dim_elements[_] for _ in range(len(dims))])}  
        
    # write cells to tm1 that have elements existing in their dimensions
    tm1.cells.write(cube_name, cellset_as_dict=cells_possible_to_write)  

    # log those cases where at least one element does not exist in the respective dimension.
    logging.warning(f'Some cells were not written to tm1. \n {cell_not_possible_to_write}')
```
